File: Week04/Mod02-destroyingarray.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll  long long int
#define vi  vector<int>
#define vll vector<long long int>
 
class UnionFind {
    private:
        vi p, rank, setSize;
        vll sumSet;
        int numSets;
    public:

        // Initialization

        UnionFind(int N){
            // We create space for N sets, but initially there are no sets.

            p.assign(N, -1);
                        
            // the height of all trees are not defined:
            rank.assign(N, -1);

            // the sizes of the individual sets are zero, as is sumset:
            setSize.assign(N, 0);
            sumSet.assign(N, 0);

            // the number of sets is N:
            numSets = 0;
        }

        void makeSet(int i, int x){
            p[i] = i;
            rank[i] = 0;
            setSize[i] = 1;
            sumSet[i] = x;
            numSets += 1;
        }

        // FindSet(i): return the representative element of the set that i belongs to.

        int findSet(int i){
            // if already at a representative element,
            // signified by the fact that the parent pointer points to the element itself,
            // return;

            if(p[i] == i)
                return i;

            // otherwise find the representative of the parent,
            // but also perform path compression by pointing the current element
            // to the final representative:

            else
                return p[i] = findSet(p[i]);
        }

        // isSameSet(i,j): returns true if and only if i and j belong to the same set.

        bool isSameSet(int i, int j){
            // Observe that i and j belong to the same set
            // if and only if they have the same representative.
            return findSet(i) == findSet(j); 
        }

        // sizeOfSet(i): returns the size of the set that the element i belongs to.

        int sizeOfSet(int i){
            return setSize[findSet(i)];
        }
        
        // numDisjointSets(): returns the number of sets being maintained currently.

        int numDisjointSets(){
            return numSets; 
        }
        
        // sumOfSet(): returns the sum of all elements of the set

        ll sumOfSet(int i){
            return sumSet[findSet(i)]; 
        }
        void unionSet(int i, int j){
            if(isSameSet(i,j)) 
                return; 
            
            int x = findSet(i);
            int y = findSet(j);

            if(rank[x] > rank[y])
                swap(x,y);
            
            p[x] = y; 

            if (rank[x] == rank[y])
                rank[y]++;
            
            setSize[y] += setSize[x];
            sumSet[y] += sumSet[x];
            numSets--;

        }


};

int main(){
    
    #ifndef ONLINE_JUDGE 
        freopen("in.txt", "r", stdin);
        freopen("out.txt", "w", stdout);
    #endif

    std::ios::sync_with_stdio(false);


    int n;
    cin >> n;
    
    // The elements of the input array.
    vi arr(n+1,0);
    
    // state[i] is 0 if position i 
    // is currently destroyed/missing, 1 otherwise.
    vi state(n+1,0);
    
    // The destruction sequence. 
    // This is a permutation of 1 to n. 
    vi sequence(n);
    
    for(ll i = 1; i <= n; i++){
        cin >> arr[i];
    }
        
    for(int i = 0; i < n; i++){
        cin >> sequence[i];
    }
 
    // Initialize the DSU data structure
    // n+1 for 1-based indexing
    UnionFind UF(n+1);

    // Reverse the sequence of deletions.
    reverse(sequence.begin(),sequence.end());
 
    // Declare an array to track the answers.
    vll answers;

    // The last answer is always zero.
    answers.push_back(0);
    
    ll currentans = 0;
    
    for(auto x: sequence){

        // The x-th index is now occupied:
        state[x] = 1;

        // Create a singleton set out of the x-th index
        // with the sum being initialized to the value arr[x]
        UF.makeSet(x,arr[x]);
        
        // Merge with the left chunk if the location
        // to the immediate left is occupied:
        // remembering to avoid the edge case:
        if(x - 1 > 0 and state[x-1])
            UF.unionSet(x-1,x);
        
        // Merge with the right chunk if the location
        // to the immediate right is occupied,
        // remembering to avoid the edge case:
        if(x + 1 <= n and state[x+1])
            UF.unionSet(x,x+1);

        // Update the answer by checking against
        // the sum of the newly created set: 
        currentans = max(UF.sumOfSet(x),currentans); 

        // Record the answer for this stage:
        answers.push_back(currentans);
    }
    
    // The last answer (the sum of the whole array)
    // is not required:

    answers.pop_back();

    // Remember to reverse the lsit of answers! 
    reverse(begin(answers),end(answers));
    
    for(auto x: answers)
        cout << x << endl;
 
}

File: Week04/Mod02-destroyingarray.py
# Problem Link: https://codeforces.com/problemset/problem/722/C

class DSU:
  def __init__(self, n):
    """Initializes the data structure spaces for 'n' sets"""
    
    # No element's parent is defined as of yet
    self.parent = [-1] * n

    # The height of all trees is not defined yet
    self.rank = [-1] * n

    # The size of each set is 0
    self.setSize = [0] * n

    # The sum of each set is also 0
    self.setSum = [0] * n

    # There are no sets initially
    self.setCount = n

  def makeSet(self, i, x):
    """Creates a new singleton set with the given 'x' as value at index 'i'"""

    # It is it's own parent, hence at rank 0
    self.parent[i] = i
    self.rank[i] = 0

    # It is a Singleton set, hence the sum is itself
    self.setSize[i] = 1
    self.setSum[i] = x

    # Increment the set count
    self.setCount += 1

  def findSet(self, i):
    """Returns the parent element of the set that 'i' belongs to"""
    
    # If the element is already a parent element of a set, return it
    if self.parent[i] == i:
      return i
    
    # Else, move up the tree to find the parent, and perform path compression
    self.parent[i] = self.findSet(self.parent[i])

    # Return the parent
    return self.parent[i]

  def sameSet(self, i, j):
    """Checks if the two elements belong to the same set"""
    return self.findSet(i) == self.findSet(j)

  def union(self, i, j):
    """Combines the sets that 'i' and 'j' belong to"""

    # if the elements belong to the same set, we don't need to do anything
    # If they're not, we perform the union operation
    if not self.sameSet(i, j):

      # Find the parent elements of the two elements
      x = self.findSet(i)
      y = self.findSet(j)

      # Make sure that 'x' is the tree with the smaller rank (proxy to depth)
      x, y = (y, x) if self.rank[x] > self.rank[y] else (x, y)

      # Set y as the parent of x
      self.parent[x] = y

      # If the ranks were the same, then the depth of the final tree must increase by 1
      if self.rank[x] == self.rank[y]:
        self.rank[y] += 1
      
      # Update the size of the combined set
      self.setSize[y] += self.setSize[x]

      # Update the sum of the combined set
      self.setSum[y] += self.setSum[x]

      # Lastly, reduce the set count, since we merged two sets into one
      self.setCount -= 1

  def sizeOfSet(self, i):
    """Returns the size of the set that the given element belongs to"""
    return self.setSize[self.findSet(i)]
  
  def getSetCount(self):
    """Returns the current count of disjoint sets"""
    return self.setCount

  def sumOfSet(self, i):
    """Returns the sum of the set that the given element belongs to"""
    return self.setSum[self.findSet(i)]

if __name__ == '__main__':
  n = int(input())
  arr = [int(i) for i in input().split()]
  sequence = [int(i)-1 for i in input().split()]

  # Reverse the sequence since we are going in reverse
  sequence.reverse()

  # To store the state of each element (1 -> not destroyed, 0 -> destroyed)
  state = [0] * n
  
  # List of answers at each stage, and a current answer variable
  answers = [0]
  currentAns = 0

  disjointSets = DSU(n)
  for idx in sequence:
    # Create the singleton set
    state[idx] = 1
    disjointSets.makeSet(idx, arr[idx])

    # Form unions with the left and right elements, if they're not destroyed
    if idx > 0 and state[idx-1]:
      disjointSets.union(idx-1, idx)
    if idx < n-1 and state[idx+1]:
      disjointSets.union(idx, idx+1)

    # Find the answer at this stage and append it to the answer array
    currentAns = max(disjointSets.sumOfSet(idx), currentAns)
    answers.append(currentAns)

  # Pop the last element and reverse the answer array and disply the answers
  answers.pop()
  answers.reverse()
  for answer in answers:
    print(answer)